server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -4111,4 +4110,26 @@ RetentionLeaseSyncer getRetentionLeaseSyncer() {
     public GatedCloseable<SegmentInfos> getSegmentInfosSnapshot() {
         return getEngine().getSegmentInfosSnapshot();
     }
+
+    /**
+     * With segment replication enabled - prepare the shard's engine to be promoted as the new primary.
+     *
+     * If this shard is currently using a replication engine, this method:
+     * 1. Invokes {@link NRTReplicationEngine#commitSegmentInfos()} to ensure the engine can be reopened as writeable from the latest refresh point.
+     * InternalEngine opens its IndexWriter from an on-disk commit point, but this replica may have recently synced from a primary's refresh point, meaning it has documents searchable in its in-memory SegmentInfos
+     * that are not part of a commit point.  This ensures that those documents are made part of a commit and do not need to be reindexed after promotion.
+     * 2. Invokes resetEngineToGlobalCheckpoint - This call performs the engine swap, opening up as a writeable engine and replays any operations in the xlog. The operations indexed from xlog here will be
+     * any ack'd writes that were not copied to this replica before promotion.
+     */
+    private void promoteNRTReplicaToPrimary() {
+        assert shardRouting.primary() && indexSettings.isSegRepEnabled();
+        getReplicationEngine().ifPresentOrElse(engine -> {
+            try {
+                engine.commitSegmentInfos();
+                resetEngineToGlobalCheckpoint();
```

Comment:
1. Can we add checks/assertions on currentGlobalCheckpoint and maxSeqNo before resetting engine to Global checkpoint
2. Can we add assertions on current routing and new routing as well

Reply:
1.  With segrep enabled we don't want to reindex ops that have already been indexed.  To avoid this the replica does not actually reset to the global cp.  In this method, when we commit we remove all other commits so the new one is always selected when opening InternalEngine. This means we allow the local cp and/or max seqNo in the commit to be higher than the global cp and not actually "safe", but I don't think we care? The use of `resetEngineToGlobalCheckpoint` here is reused to swap the engine type, and potentially replay up to the global cp.

2. Added this before this method is invoked inside `updateShardState`.